// eslint-plugin-wc/lib/rules/no-customized-built-in-elements.js
"use strict";
/**
 * @fileoverview Disallows extending of built-in elements
 * @author James Garbutt <https://github.com/43081j>
 * @author Keith Cirkel <https://github.com/keithamus>
 */
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../util");
const tag_names_1 = require("../util/tag-names");
//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------
const rule = {
    meta: {
        docs: {
            description: 'Disallows extending of built-in elements',
            url: 'https://github.com/43081j/eslint-plugin-wc/blob/master/docs/rules/no-customized-built-in-elements.md'
        },
        messages: {
            noBuiltIn: 'You should avoid extending built-in elements as support ' +
                'is not guaranteed.'
        }
    },
    create(context) {
        // variables should be defined here
        const source = context.sourceCode;
        //----------------------------------------------------------------------
        // Helpers
        //----------------------------------------------------------------------
        //----------------------------------------------------------------------
        // Public
        //----------------------------------------------------------------------
        return {
            'ClassDeclaration,ClassExpression': (node) => {
                if ((0, util_1.isCustomElement)(context, node, source.getJSDocComment(node)) &&
                    node.superClass &&
                    node.superClass.type === 'Identifier' &&
                    node.superClass.name !== 'HTMLElement' &&
                    tag_names_1.builtInTagClassList.includes(node.superClass.name)) {
                    context.report({
                        node,
                        messageId: 'noBuiltIn'
                    });
                }
            }
        };
    }
};
exports.default = rule;
