// eslint-plugin-wc/lib/rules/no-method-prefixed-with-on.js
/**
 * @fileoverview Disallows methods prefixed with `on`
 * @author James Garbutt <https://github.com/43081j>
 * @author Keith Cirkel <https://github.com/keithamus>
 */
import { isCustomElement } from '../util.js';
import { getMethodName } from '../util/ast.js';
//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------
const rule = {
    meta: {
        docs: {
            description: 'Disallows methods prefixed with `on`',
            url: 'https://github.com/43081j/eslint-plugin-wc/blob/master/docs/rules/no-method-prefixed-with-on.md'
        },
        messages: {
            noPrefix: 'Avoid using `on` as a prefix of method names as they can ' +
                'easily conflict with reserved event handler names'
        }
    },
    create(context) {
        // variables should be defined here
        let insideElement = false;
        const source = context.sourceCode;
        //----------------------------------------------------------------------
        // Helpers
        //----------------------------------------------------------------------
        //----------------------------------------------------------------------
        // Public
        //----------------------------------------------------------------------
        return {
            'ClassDeclaration,ClassExpression': (node) => {
                if (isCustomElement(context, node, source.getCommentsBefore(node))) {
                    insideElement = true;
                }
            },
            'ClassDeclaration,ClassExpression:exit': () => {
                insideElement = false;
            },
            MethodDefinition: (node) => {
                if (insideElement) {
                    const name = getMethodName(node);
                    if (name?.startsWith('on')) {
                        context.report({
                            node,
                            messageId: 'noPrefix'
                        });
                    }
                }
            }
        };
    }
};
export default rule;
